### PR TITLE
Support header and struct for logging (fixes #2201 from p4c)

### PR DESCRIPTION
**Pull request by lcikaric**
This PR is related to p4lang/p4c#2596 PR in p4c.

Example 1:
header Hdr {
	int<4> a;
	bit<4> b;
}
Hdr h = {3,5};
log_msg("h = {}", {h});
Output: h = {'a': 3, 'b': 5, '$valid$': 1}

Example 2:
struct Str {
	int<4> a;
	bit<4> b;
	bool c;
}
Str s = {3,5,true};
log_msg("s = {}", {s});
Output: s = {'a': 3, 'b': 5, 'c': 1}

### CODE REVIEW

**Comment by jafingerhut** [comment]
Looks like a useful enhancement to me.

Minor suggestion: For headers, it might be nice if the value of $valid$ was always first in the output, before the other fields.  It might be reasonable to log invalid header values as `{'$valid$':0}` with no other fields shown, but I do not have a strong opinion on that as long as it is clear from the output when a header is valid vs. invalid.

I suspect taking advantage of these enhancements, if added to behavioral-model, will require some changes to p4c bmv2 back end code that might give an error, but of course those changes are not worth making until and unless these changes are approved and merged in.

File: src/bm_sim/P4Objects.cpp [hunk 1]
```diff
@@ -336,7 +336,32 @@ P4Objects::process_single_param(ActionFn* action_fn,
...
+          } else if (type_cfg == "bool") {
+              int val = 0;
+              if (cfg["value"].asString() == "true") {
+                val = 1;
+              }
+              vec_data.push_back(std::make_pair(position, Data(val)));
+          }
+          position++;
+        }
+        action_fn->parameter_push_back_list(std::unique_ptr<List>(new List(vec_data, vec_header_id)));
```
Comment (antoninbas) [inline src/bm_sim/P4Objects.cpp:361]: I don't think the generic notion of `List` should be introduced at all, at least not in this PR. bmv2 purposely only deals with headers (named structs can be mapped to headers) and is not aware of tuples. We should not change that just for `log_msg`.

File: include/bm/bm_sim/data.h [hunk 2]
```diff
@@ -38,6 +38,17 @@ namespace bm {
 
 using bignum::Bignum;
 
+// DataRepresentation is used to represent data for log_msg.
+// if log_msg is to support a data type, the class for that type
+// should inherit this class and override show method.
+// For now the subclasses are Data, Header and List.
+class DataRepresentation {
```
Comment (antoninbas) [inline include/bm/bm_sim/data.h:45]: I think the convention in the rest of the code is to use the `Iface` suffix for pure abstract classes. So I would suggest `StringRepresentationIface` and maybe the `show` method should be renamed to `get_string_repr`. There is already a `get_string_repr` method in `Data` and your `show` implementation is calling it so why not use it directly?

File: include/bm/bm_sim/headers.h [hunk 3]
```diff
@@ -280,6 +282,35 @@ class Header : public NamedP4Object {
   //! string. The name is of the form <hdr_name>.<f_name>.
   const std::string get_field_full_name(int field_offset) const;
 
+  // The header and structure are represented by the class Header in bmv2.
+  // The difference between the header and the structure is in the metadata variable:
+  //  metadata = 0 for the header
+  //  metadata = 1 for the structure.
+  std::string show() const override {
```
Comment (antoninbas) [inline include/bm/bm_sim/headers.h:289]: move to cpp file

File: include/bm/bm_sim/headers.h [hunk 4]
```diff
@@ -280,6 +282,35 @@ class Header : public NamedP4Object {
...
+
+    if (!is_metadata() && !is_valid()) {
+      std::string name = get_name();
+      return "Header " + name.substr(0,name.rfind('_')) + " is not valid";
+    }
```
Comment (antoninbas) [inline include/bm/bm_sim/headers.h:294]: I think the code should be more uniform and look like this:
 * if it's a header (i.e. `metadata` is false), print validity information first
 * then print all the *non-hidden* fields, so all fields up to the hidden offset

This special case is not needed IMO: even for a non-valid header, you should still honor the P4 program and print all field values

File: include/bm/bm_sim/actions.h [hunk 5]
```diff
@@ -521,24 +523,51 @@ const char *ActionParam::to<const char *>(ActionEngineState *state) const {
...
+// only support "const std::vector<const DataRepresentation*>" as the type used by the C++ primitive
 // and currently this is only used by log_msg.
 // TODO(antonin): more types (e.g. "const std::vector<const Data &>") if needed
 // we can support list of other value types as well, e.g. list of headers
 // we are not limited to using a vector as the data structure either
 template <> inline
-const std::vector<Data>
-ActionParam::to<const std::vector<Data>>(ActionEngineState *state) const {
+const std::vector<const DataRepresentation*>
+ActionParam::to<const std::vector<const DataRepresentation*>>(ActionEngineState *state) const {
```
Comment (antoninbas) [inline include/bm/bm_sim/actions.h:552]: keep the current function and add a new one for your use case

**Review by antoninbas — CHANGES_REQUESTED** [review]
My main comment is that I do not see a compelling reason to introduce the notion of list and I believe it should be removed. bmv2 expects all headers to be flattened (structures are headers) and does not understand tuples. A "parameter vector" is already a list of things where you can mix integral values and headers. This covers the examples you include in the PR description and the ones called out in https://github.com/p4lang/p4c/issues/2201. I suggest not trying to cover tuples in this PR. It should be done in a subsequent PR and done properly.

A few additional comments:
 * the JSON format doc (https://github.com/p4lang/behavioral-model/blob/master/docs/JSON_format.md) probably needs to be updated (the part about `log_msg`)
 * I don't see any unit tests

File: src/bm_sim/P4Objects.cpp (hunk 1 above)
Comment (lcikaric) [reply src/bm_sim/P4Objects.cpp:361]: The main idea of this fix is support for log_msg for headers and structures. Structures which contain header as a field are represented same as lists and tuples in JSON file, so if we want to support them, we will support lists and tuples too.

File: src/bm_sim/P4Objects.cpp [hunk 6]
```diff
@@ -336,7 +336,32 @@ P4Objects::process_single_param(ActionFn* action_fn,
...
+            const auto header_name = cfg["value"].asString();
+            auto header_id = get_header_id_cfg(header_name);
+            vec_header_id.push_back(std::make_pair(position, header_id));
+          } else if (type_cfg == "bool") {
+              int val = 0;
+              if (cfg["value"].asString() == "true") {
+                val = 1;
+              }
+              vec_data.push_back(std::make_pair(position, Data(val)));
+          }
```
Comment (antoninbas) [inline src/bm_sim/P4Objects.cpp:358]: this could be other things, like a `"field"`

File: src/bm_sim/P4Objects.cpp (hunk 1 above)
Comment (antoninbas) [reply src/bm_sim/P4Objects.cpp:361]: There are 2 issues there IMO:
* it doesn't respect the type-value syntax used in the JSON format (https://github.com/p4lang/behavioral-model/blob/master/docs/JSON_format.md#the-type-value-object), some new `type` should be introduced
* there is some kind of asymmetry as inner-most structs are treated as headers, whereas outer structs are JSON arrays?

File: include/bm/bm_sim/headers.h [hunk 7]
```diff
@@ -320,6 +351,88 @@ class Header : public NamedP4Object {
 #endif
 };
 
+
+// This is a helper class.
+// For now, this class is only used to represent the list, tuple and
+// structure with headers required for log_msg.
+class List : public DataRepresentation {
```
Comment (antoninbas) [inline include/bm/bm_sim/headers.h:358]: is the order not respected since multiple vectors are being used?

**Comment by lcikaric** [comment]
> There are 2 issues there IMO:
> 
>     * it doesn't respect the type-value syntax used in the JSON format (https://github.com/p4lang/behavioral-model/blob/master/docs/JSON_format.md#the-type-value-object), some new `type` should be introduced
> 
>     * there is some kind of asymmetry as inner-most structs are treated as headers, whereas outer structs are JSON arrays?

[test.p4.txt](https://github.com/p4lang/behavioral-model/files/5480971/test.p4.txt)
[out.json.txt](https://github.com/p4lang/behavioral-model/files/5480972/out.json.txt)

As you can see from provided p4 test and p4c JSON output, structure 'str1' which does not contain header as a field is represented as a header and structure 'str2' which contains header 'hdr1' is represented as a json array in value of type "parameters_vector" for log_msg. So there is asymmetry in representing inner-most struct and outer ones as you said.

Have you meant to introduce a new type for representing outer structs which have structs and headers as a fields? It makes sense to me.

File: include/bm/bm_sim/data.h (hunk 2 above)
Comment (lcikaric) [reply include/bm/bm_sim/data.h:45]: I've implemented the change you suggested and I've fixed format issues. I will upload the code once I do further analysis based on the comments that follow and when I add unit test.

File: include/bm/bm_sim/headers.h (hunk 7 above)
Comment (lcikaric) [reply include/bm/bm_sim/headers.h:358]: If you mean the order of the list fields, it is respected when logging message.

**Comment by lcikaric** [comment]
I made changes according to your suggestions and I updated pr. What should be my further steps in order to merge this pr?

**Comment by lcikaric** [comment]
@antoninbas Can you suggest next steps on pr, please?

**Comment by antoninbas** [comment]
Very sorry about the delay, I missed the first notification (for https://github.com/p4lang/behavioral-model/pull/962#issuecomment-750922152).

Looking at this PR again, I think most of my original comments still stand. If we take your example:
```p4
Header hdr1 = {4,true};
Structure2 str2 = {3,hdr1};
log_msg("str2 = {}",{str2});
```
and the corresponding section from the generate JSON:
```json
        {
          "op" : "log_msg",
          "parameters" : [
            {
              "type" : "string",
              "value" : "str2 = {}"
            },
            {
              "type" : "parameters_vector",
              "value" : [
                [
                  {
                    "type" : "hexstr",
                    "value" : "0x03"
                  },
                  {
                    "type" : "header",
                    "value" : "hdr1_0"
                  }
                ]
              ]
            }
          ],
          "source_info" : {
            "filename" : "test.p4",
            "line" : 80,
            "column" : 8,
            "source_fragment" : "        log_msg(\\\"str2 = {}\\\",{str2});"
          }
        }
```

As far as I'm concerned, this JSON is now what we want. It doesn't include any information about `str2` (field names).
The JSON should look like this:
```json
{
              "type" : "parameters_vector",
              "value" : [
                {
                  "type" : "header",
                  "value" : "str2_0"
                }
              ]
            }
```
Where `str2_0` is an instance of type `Structure2`, in which all fields have been flattened:
```json
    {
      "name" : "Structure2",
      "id" : 2,
      "fields" : [
        ["val1", 7, true],
        ["val2.val1", 7, false],
        ["val2.val2", 1, false],
        ["_padding", 1, false]
      ]
    }
```
In other words, I would expect symmetry with `str1`. And for that example, I don't see any value to the concept of `List`.

**Comment by lcikaric** [comment]
Thank you for your response. I agree with you. I'm working on it.
